### PR TITLE
report 100% success rate in case there is no new data

### PR DESCRIPTION
**Pull request by darkdarkdragon**
**What does this pull request do? Explain your changes. (required)**
report 100% success rate in case there is no running streams
(this is needed because Prometheus will continue to show last value
if there is no new data)

**Specific updates (required)**

- 
- 
- 

**How did you test each of these updates (required)**
Manually

**Does this pull request close any open issues?**

**Checklist:**

- [ ] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

### CODE REVIEW

**Comment by j0sh** [comment]
Looks OK. Reporting "100% success" feels kind of strange if there isn't actual work being done. Is there some sort of intermediate state such as "inactive" that we could report? I'm concerned that this may distort the metrics over time, eg

T1. Success (active)
T2. Success (inactive)
T3. Fail (active)

Here the metrics would indicate a 66% success rate from T1-T3 when it's actually 50%.

**Comment by darkdarkdragon** [comment]
This is needed because Prometheus client library doesn't have ability to remove metric from being reported. 

> Here the metrics would indicate a 66% success rate from T1-T3 when it's actually 50%.

With current algo this situation isn't possible - once stream ended, it gets removed from success rate calculation immediately. But this means we could loose failures of up to four last segments (if they not time fail immediately, but, say, fail because of timeout).
Do you think I should improve algo to account for this possibility?

**Comment by j0sh** [comment]
> Do you think I should improve algo to account for this possibility?

Maybe - perhaps I misunderstand but how does the success rate metric / algorithm work right now?

**Comment by darkdarkdragon** [comment]
for every stream there is ring buffer (`segmentsAverager`) which keeps last 30 segments emerged. For each segment I'm recording `seqNo`, time in which it appeared and emerged/transcoded flags. Overall success rate is average of success rate of every stream. Stream's success rate is number of successfully transcoded segments divided by number of emerged segments (excluding segments that was emerged less than 8.5s ago and hasn't received success or failure event) - so success rate for stream is essentially is moving average with 60s window, and timeout errors gets delayed by 8.5 seconds.
But after stream ends, whole `segmentsAverager` structure with list of segments gets removed, so any possible timeout errors for last 8.5s gets lost.
After second thought, I'll change algorithm to wait for last segments completion or timeout, will do this with another PR.

**Comment by darkdarkdragon** [comment]
@j0sh Now if last couple of segments timeout, it will be reflected in success rate.
Also I've added unit tests.

File: server/broadcast.go [hunk 1]
```diff
@@ -236,7 +238,7 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) {
 		}
 		cond.L.Unlock()
 		if monitor.Enabled {
-			monitor.SegmentFullyTranscoded(nonce, seg.SeqNo, common.ProfilesNames(sess.Profiles), len(segHashes) == len(res.Segments))
+			monitor.SegmentFullyTranscoded(nonce, seg.SeqNo, common.ProfilesNames(sess.Profiles), len(segHashes) == len(res.Segments), errCode)
```
Comment (j0sh) [inline server/broadcast.go]: Just noting that we'll exit early if there's an error with the download (in order to retry), so I hope we aren't using the SegmentFullyTranscoded call to wrap up any stateful tracking:  https://github.com/livepeer/go-livepeer/commit/2dcf826485969be9415cc0b93b2237d867ecc751

The params here are getting sort of long, would just passing in the error suffice to determine whether `len(segHashes) == len(res.Segments)` ? Does the `allSuccess` parameter convey any additional data?

Additionally, we might want to incorporate any errors from a failed sig check.

Comment (darkdarkdragon) [reply server/broadcast.go]: > I hope we aren't using the SegmentFullyTranscoded call to wrap up any stateful tracking

We are. Adjust metrics to retries is my next task.

> would just passing in the error suffice to determine whether len(segHashes) == len(res.Segments)

Changed it to use `gotErr` flag.

> Additionally, we might want to incorporate any errors from a failed sig check.

Will do this later, when we will start doing with this information something else than just printing it to log.

**Review by j0sh — APPROVED** [review]
LGTM! 🚢

**Pull request merged**